src/org/oidc/service/base/ServiceContext.java

```diff
@@ -0,0 +1,328 @@
+     */
+    private String clientId;
+    /**
+     * The client secret, which is optional.
+     */
+    private String clientSecret;
+    /**
+     * The Issuer ID. This is the unique identifier of the OP/AS the
+     * client is communicating with.
+     */
+    private String issuer;
+    /**
+     * An unordered list of redirect uris that the client expects to use.
+     */
+    private List<String> redirectUris;
+    /**
+     * redirectUris contains a list of unspecified redirect uris.
+     * In reality there are good reasons for having separated redirect
+     * uris for different expected response modes.
+     * At this time we know of 3 such modes:
+     * fragment, queryString, formPost
+     * In callback, we can keep the redirect uris per response mode
+     * separate.
+     */
+    private Map<DataLocation,String> callBack;
+
+    /**
+     * Constants
+     */
+    private static final String SIG = "sig";
+    private static final String RSA = "rsa";
+    private static final String SHA_256 = "SHA-256";
+    private static final String ISSUER = "issuer";
+
+    public ServiceContext(
+            KeyJar keyJar,
+            ServiceContextConfig config) {
+        this.keyJar = keyJar;
+        this.config = config;
+    }
+
+    public ServiceContext() {
+
+    }
+
+    /**
+     The client needs its own set of keys. It can either dynamically create them or load them from local storage. This method can also fetch other entities keys provided that the URL points to a JWKS.
+     * @param keySpecifications contains fileName and algorithm
+     **/
+        public void importKeys(Map<FileOrUrl,KeySpecifications> keySpecifications) {
+        if(keySpecifications == null) {
+            throw new IllegalArgumentException("null keySpecifications");
+        }
+
+        Set<FileOrUrl> keys = keySpecifications.keySet();
+        KeySpecifications keySpecificationsIndex;
+        Key rsaKey;
+        KeyBundle keyBundle;
+        for(FileOrUrl key : keys) {
+            if(FileOrUrl.FILE.equals(key)) {
+                keySpecificationsIndex = keySpecifications.get(key);
+                if(RSA.equalsIgnoreCase(keySpecificationsIndex.getAlgorithm())) {
+                    rsaKey = new RSAKey(Jwk.importPrivateRsaKeyFromFile(keySpecificationsIndex.getFileName()), SIG);
+                    keyBundle = new KeyBundle();
+                    keyBundle.addKey(rsaKey);
+                    keyJar.addKeyBundle("", keyBundle);
+                }
+            } else if (FileOrUrl.URL.equals(key)) {
+                keyBundle = new KeyBundle();
+                keyJar.addKeyBundle(ISSUER,keyBundle);
+            }
+        }
+    }
+
+    /**
+     A 1<->1 map is maintained between a URL pointing to a file and
+     the name of the file in the file system.
+
+     As an example if the base_url is 'https://example.com' and a jwks_uri
+     is 'https://example.com/jwks_uri.json' then the filename of the
```

Comment:
This is a happy path url. While unit test, please consider - different cases, special characters etc.